lib/kubernetes-deploy/runner_task.rb

```diff
@@ -15,43 +15,115 @@ def initialize(task_template, namespace, context)
+      if error.message != error.class.to_s
+        @logger.summary.add_action(error.message)
+        @logger.print_summary(:failure)
+      end
       false
     end
 
-    def run!(task_template:, entrypoint:, args:, env_vars: [])
+    def run!(task_template:, entrypoint:, args:, env_vars: [], verify_result: true)
       @logger.reset
-      @logger.phase_heading("Validating configuration")
+      @logger.phase_heading("Initializing deploy")
       validate_configuration(task_template, args)
       if kubectl.server_version < Gem::Version.new(MIN_KUBE_VERSION)
         @logger.warn(KubernetesDeploy::Errors.server_version_warning(kubectl.server_version))
       end
-      @logger.phase_heading("Fetching task template")
       raw_template = get_template(task_template)
 
-      @logger.phase_heading("Constructing final pod specification")
       rendered_template = build_pod_template(raw_template, entrypoint, args, env_vars)
-
-      validate_pod_spec(rendered_template)
-
+      validate_restart_policy(rendered_template, verify_result)
+      pod = Pod.new(namespace: @namespace, context: @context, logger: @logger, log_on_success: false,
+                    definition: rendered_template.to_hash.deep_stringify_keys, statsd_tags: [])
+      pod.validate_definition(kubectl)
+      @logger.info("Configuration valid")
       @logger.phase_heading("Creating pod")
       @logger.info("Starting task runner pod: '#{rendered_template.metadata.name}'")
+
+      pod_creation_time = Time.now.utc
+      pod.deploy_started_at = pod_creation_time
       @kubeclient.create_pod(rendered_template)
+
+      if verify_result
+        sync_mediator = SyncMediator.new(namespace: @namespace, context: @context, logger: @logger)
+        watch_and_report(pod, pod_creation_time, sync_mediator)
+      else
+        warning = <<~MSG
+          Result verification is disabled for this task.
+          This means the desired pod was successfully created, but the runner did not make sure it actually succeeded.
+        MSG
+        @logger.summary.add_paragraph(ColorizedString.new(warning).yellow)
+      end
     end
 
     private
 
+    def watch_and_report(pod, deploy_started_at, sync_mediator)
+      sync_interval = 5.seconds
+      last_loop_time = last_log_time = deploy_started_at
+      @logger.info("Logs from #{pod.id}:")
+
+      loop do
+        pod.sync(sync_mediator)
+        logs = pod.fetch_logs(kubectl, since: last_log_time.to_datetime.rfc3339)
+        last_log_time = Time.now.utc
+
+        logs.each do |_, log|
+          if log.present?
+            @logger.info("\t" + log.join("\n\t"))
+          else
+            @logger.info("\t...")
+          end
+        end
+
+        if (sleep_duration = sync_interval - (Time.now.utc - last_loop_time)) > 0
+          sleep(sleep_duration)
+        end
+        last_loop_time = Time.now.utc
+
+        return if deploy_ended?(pod, deploy_started_at)
+      end
+    end
+
+    def deploy_ended?(pod, deploy_started_at)
```

Comment:
I dislike that this predicate method (`?`) has so many side-effects (it logs and it might raise an exception). That's unconventional. In Ruby, predicate-methods should either return `true` or `false` but not do anything else.

Reply:
I feel like `deploy_ended?` should be a method on `Pod`, not a method here. And it shouldn't do anything related to timeouts but rather that should be a separate method (`deploy_timed_out?` or `raise_if_deploy_timed_out`).

Reply:
`watch_and_report` has been refactored into its own class, also fixed up this method since it lives there now. `deploy_ended?` we know we have an issue with this and there is a seperate PR investigating better state tracking.